STREAM-38 Implement PIPE redefinition

akuma432 opened the pull request:
* added snowpipe-redefinition.sql.ssp template for pipe redefinition if schema changes
* updated junit

akuma432 commented on src/main/resources/templates/snowflake/snowflake-schema-evolution/snowpipe-redefinition.sql.ssp:
```diff
@@ -0,0 +1,69 @@
+#{
+    val destination = configuration.destination.asInstanceOf[Snowflake]
+    val sourceColumnStr = table.sourceColumnConversion(typeMapping)
+}#
```
should I put a check here tableDiff.allChangesAreCompatible() ?
Because even for modify columns also these scripts are generated. 
Currently we are only processing add columns and for other operations throwing error message.

brockn replied:
Seems like yes we would only want to output this when we are doing a all changes are compatible.

akuma432 replied:
updated

akuma432 commented on src/test/resources/results/scriptCommand/testScriptCommandForSchemaEvolution_Alter_Table_Modify_column/Employee/snowpipe-redefinition.sql:
```diff
@@ -0,0 +1,60 @@
+USE DATABASE SANDBOX_POC1;
```
Since this is case of modify columns and we are ignoring these changes for now I think we should neither generate script folder nor any sql scripts for these tables.

akuma432 commented on build.sbt:
```diff
@@ -118,6 +118,13 @@ mappings in Universal ++= {
+      f -> s"templates/snowflake/snowflake-schema-evolution/${f.name}"
+    }
+}
+
```
to add snowflake-schema-evolution templates during packaging

brockn commented:
I am wondering if:

1. We shouldn't rename evolve-schema.ssp to create-table-evolve-schema.ssp.
2. And pipe-redefinition to create-pipe-evolve-schema.ssp and have it do the initial creation of the pipe unless there are changes in which case it invoke this logic. Feels cleaner.

akuma432 commented:
> I am wondering if:
> 
> 1. We shouldn't rename evolve-schema.ssp to create-table-evolve-schema.ssp.
> 2. And pipe-redefinition to create-pipe-evolve-schema.ssp and have it do the initial creation of the pipe unless there are changes in which case it invoke this logic. Feels cleaner.

1. Renamed evolve-schema.ssp to create-table-evolve-schema.ssp.
2. Renamed snowpipe-redefinition to create-pipe-evolve-schema.ssp. create-pipe-evolve-schema.ssp will be used only during evolve schema and here we have following use case :
  a. add column -> new column is added hence pipe redefinition 
  b. modify column -> let's say data type is changed or length is changed. I think here also we should redefine pipe.
  c. delete column -> a column is deleted hence pipe redefinition
  d. table created -> a new table is created hence initial pipe creation is needed.
  e. table dropped -> drop the pipe. 

For now I am implementing use case A and D. For others no query is generated. According to above use case I have modified the script.    
 Please let me know your feedback on this.

akuma432 commented on src/main/resources/templates/snowflake/shared/create-snowpipe.ssp:
```diff
@@ -0,0 +1,22 @@
+#import(io.phdata.streamliner.schemadefiner.model.Snowflake)
```
create-snowpipe is used in two templates to added it in shared folder and used in others.

akuma432 commented on src/main/resources/templates/snowflake/snowflake-schema-evolution/create-pipe-evolve-schema.sql.ssp:
```diff
@@ -0,0 +1,79 @@
+        DROP PROCEDURE streamliner_recreate_pipe(string);
+    #end
+#elseif (tableDiff.existsInSource && !tableDiff.existsInDestination)
+    <% render("../shared/create-snowpipe.ssp", Map("configuration" -> configuration, "table" -> table, "typeMapping" -> typeMapping))%>
```
else block is for create table hence initial pipe creation will happen.

akuma432 commented on src/test/resources/results/scriptCommand/testScriptCommandForSchemaEvolution_Create_Table/Department/create-pipe-evolve-schema.sql:
```diff
@@ -0,0 +1,11 @@
+USE SCHEMA EMPLOYEES;
+
+CREATE PIPE IF NOT EXISTS Department_pipe
+	AUTO_INGEST = true
```
this is use case of schema evolution where a new table is created. Hence initial pipe is created.
